client/src/main/scala/org/http4s/client/middleware/Retry.scala

```diff
@@ -2,36 +2,55 @@ package org.http4s
 package client
 package middleware
 
-import org.http4s.Status.ResponseClass.Successful
-import org.http4s.{Response, Request, EmptyBody}
-import org.http4s.client.Client
-import org.log4s.getLogger
-
 import scala.concurrent.duration._
 import scala.math.{pow, min, random}
-import scalaz.Kleisli
 
+import scalaz._
 import scalaz.concurrent.Task
+import org.http4s.Status._
+import org.log4s.getLogger
 
 object Retry {
 
   private[this] val logger = getLogger
 
+  private[this] val RetriableStatuses = Set(
+    RequestTimeout,
+    // TODO Leaving PayloadTooLarge out until we model Retry-After
+    InternalServerError,
+    ServiceUnavailable,
+    BadGateway,
+    GatewayTimeout
+  )
+
   def apply(backoff: Int => Option[FiniteDuration])(client: Client) = {
     def prepareLoop(req: Request, attempts: Int): Task[DisposableResponse] = {
-      client.open(req) flatMap {
-        case dr @ DisposableResponse(Successful(resp), _) =>
+      client.open(req).attempt flatMap {
+        case \/-(dr @ DisposableResponse(Response(status, _, _, _, _), _)) if req.isIdempotent && RetriableStatuses(status) =>
+          logger.info(s"Request ${req} has failed attempt ${attempts} with reason ${status}. Retrying.")
```

Comment:
We should move these log.info messages that say its going to retry to after the `Some(duration)` to avoid a confusing log message should the `backoff` function give us a `None` and we don't actually retry.